### PR TITLE
[Common, PyTorch] Improve mHC to match DeepSeek's implementation

### PR DESCRIPTION
**Pull request by kainzhong**
# Description

Some enhancement for mHC to better align with DeepSeek's tilelang implementation: https://github.com/deepseek-ai/TileKernels/tree/main/tile_kernels/mhc

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes
- Add `mhc_generate_mix_and_aggregate` API that does projection, scale, sinkhorn and aggregate together
- Allow `mhc_fused_projection` to accept arguments with mixed dtype: x.dtype=bf16, phi.dtype=fp32, which matches DeepSeek's [implementation](https://github.com/deepseek-ai/TileKernels/blob/36d9e45d38e204ebb87e6f6e833821eee0482fe5/tile_kernels/modeling/mhc/ops/norm_fn.py#L58)
- `mhc_fused_projection` now outputs fp32 regardless of the input dtype, matching DeepSeek's [implementation](https://github.com/deepseek-ai/TileKernels/blob/36d9e45d38e204ebb87e6f6e833821eee0482fe5/tile_kernels/modeling/mhc/ops/norm_fn.py#L70)
- Add `fuse_grad_x_acc` optimization (default to False), which will reuse the same grad_x buffer to accumulate the initial mHC input x's gradient for `mhc_fused_expand_combine`, `mhc_fused_aggregate` and `mhc_fused_projection` 
- Support `norm_weight` for `mhc_fused_projection`, which would be equivalent to apply RMSNorm in the unfused manner with `elementwise_affine=True`, which would be the learnable per-element affine parameters for RMSNorm
- Support deterministic mode with non-spilt-K
- Refactor some kernel code to avoid duplication. I just realized if you make a triton kernel constexpr, it can be used as a macro in if branches since triton will not compile if it knows in compile time that some branch will not be taken
- Fix the bug that grid will exceed CUDA's limitation when M is too large and the autotune candidate is BLOCK_SIZE_M=1. Such invalid configs will be pruned now.
- Improve projection op by using TMA if on Hopper+

- ~~Add `main_grad` optimization for Megatron-LM integration, which will accumulate the gradient of phi, alpha, beta (they are all supposed to be `torch.nn.Parameter`) to `main_grad` if such attribute exists.~~ Removed because it's too much complexity and this parameter is not too large so this optimization only leads to negligible win
- ~~[TODO]: add checkpoint recomputing fused kernel to match DeepSeek's [implementation](https://github.com/deepseek-ai/TileKernels/blob/main/tile_kernels/modeling/mhc/ops/multilayer_recompute.py)~~ Probably need to figure out how to implement this in Megatron first since it doesn't seem too straightforward
- ~~[TODO]: add a fused projection + aggregate only kernel (no expand & combine path) for the last mHC layer, which seems to be also used for MTP. See function `learned_output_contract` in https://github.com/NVIDIA/Megatron-LM/pull/4518~~ That kernel only takes 0.06ms for fwd+bwd. Plus my vibe coded kernel doesn't save too much time here.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/triton/mhc.py [hunk 1]
```diff
@@ -803,7 +1018,7 @@ class mHCExpandCombineOp(torch.autograd.Function):
     """
 
     @staticmethod
-    def forward(ctx, f, bias, H_post, x, H_res, n, use_tf32=True):
+    def forward(ctx, f, bias, H_post, x, H_res, n, use_tf32=True, fuse_grad_x_acc=True):
```
Comment (kainzhong) [inline transformer_engine/pytorch/triton/mhc.py]: Make it default to False, and also add this to other two functions to make users aware

File: transformer_engine/pytorch/triton/mhc.py [hunk 2]
```diff
@@ -996,4 +1213,9 @@ def backward(ctx, grad_output):
         if bias is not None:
             grad_bias = grad_bias.to(bias.dtype)
 
-        return grad_f, grad_bias, grad_H_post, grad_x, grad_H_res, None, None
+        if ctx.fuse_grad_x_acc:
+            # When fused x gradient accumulation is enabled, use fp32 for the accumulation buffer
+            x.untyped_storage().grad_x_acc = grad_x.to(torch.float32)
```
Comment (kainzhong) [inline transformer_engine/pytorch/triton/mhc.py:1269]: Check if this is overwriting anything that is already there

File: transformer_engine/pytorch/triton/mhc.py [hunk 3]
```diff
@@ -763,7 +970,11 @@ def backward(ctx, grad_output):
         assert n == 4, "Only n=4 is supported in this implementation"
         M = s * b
 
-        grad_x = torch.empty_like(x)
+        fuse_grad_x_acc = hasattr(x.untyped_storage(), "grad_x_acc")
```
Comment (kainzhong) [inline transformer_engine/pytorch/triton/mhc.py]: Add fuse_grad_x_acc param in other functions as well

File: transformer_engine/pytorch/triton/mhc.py [hunk 4]
```diff
@@ -401,6 +614,8 @@ def forward(ctx, H, alpha, beta, ms, n):
 
         ctx.dtype = H.dtype
         H = H.to(torch.float32)
+        ctx.alpha_main_grad = getattr(alpha, "main_grad", None)
```
Comment (kainzhong) [inline transformer_engine/pytorch/triton/mhc.py]: With FSDP this will break
Check linear.py for reference
Talk with Varun

Comment (kainzhong) [reply transformer_engine/pytorch/triton/mhc.py]: After some discussion we decide to remove this `main_grad` in my PR since it's a lot of complexity but it only saves a little overhead since the parameter here is small: (4*hidden, 32).
Will probably wait until Varun figures out how to make `main_grad` nicer in his PR
